forui/lib/src/widgets/pagination/pagination_controller.dart

```diff
@@ -0,0 +1,206 @@
+
+  void _handleScrollChanges() {
+    if (hasClients) {
+      if (_controller.page case final page when page!.round() != _page) {
+        _page = page.round();
+
+        notifyListeners();
+      }
+    }
+  }
+
+  /// Returns whether the controller has clients (i.e., it is attached to a [PageView]),
+  bool get hasClients => _controller.hasClients;
+
+  /// Returns the [PageController] instance.
+  PageController get controller => _controller;
+
+  /// Returns the total number of pages in the pagination.
+  ///
+  /// If the controller has clients (i.e., it is attached to a [PageView]),
+  /// the length is dynamically computed using the scroll extents and viewport dimensions.
+  ///
+  /// If not attached to any clients, it returns the predefined `_length`, which defaults to 0.
+  num get pages {
+    if (hasClients && _controller.position.hasContentDimensions) {
+      final position = _controller.position;
+      final len =
+          max(0, position.maxScrollExtent - position.minScrollExtent) /
+          max(1.0, position.viewportDimension * _controller.viewportFraction);
+      if (len.isFinite) {
+        return len.ceil() + 1;
+      } else {
+        return len;
+      }
+    } else {
+      return _pages;
+    }
+  }
+
+  /// Moves to the previous page if the current page is greater than 1.
+  void previous() {
+    final value = hasClients && _controller.page != null ? _controller.page!.round() : _page;
+    if (0 < value) {
+      if (hasClients) {
+        _controller.previousPage(duration: const Duration(milliseconds: 300), curve: Curves.ease);
+      } else {
+        _page = value - 1;
+        notifyListeners();
+      }
+    }
+  }
+
+  /// Moves to the next page if the current page is less than the total number of pages.
+  void next() {
+    final value = hasClients && _controller.page != null ? _controller.page!.round() : _page;
+    if (value < pages - 1) {
+      if (hasClients) {
+        _controller.nextPage(duration: const Duration(milliseconds: 300), curve: Curves.ease);
+      } else {
+        _page = value + 1;
+        notifyListeners();
+      }
+    }
+  }
+
+  /// Returns the current page index.
+  int get page => _page;
+
+  /// Sets the current page index.
+  set page(int index) {
+    if (0 <= index && index <= pages) {
+      if (hasClients) {
+        _controller.jumpToPage(index);
+        _page = index;
+      } else {
+        _page = index;
+        notifyListeners();
+      }
+    } else {
+      throw StateError('The index must be within the allowed range.');
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  /// Calculates the range of page numbers to display around the current page.
+  ///
+  /// Returns a tuple of the start and end page numbers by centering the range around the current page,
+  /// ensuring a balanced display of sibling pages on either side while considering pagination constraints.
+  (int, int) calculateSiblingRange() {
```

Comment:
I think this can be a getter instead. (Remember to update the docs when changing). Something like `The start and end pages to display around the current page`.